RSA Signing & Verification Support

liamdennehy opened the pull request:
RSA signing is largely similar to hmac, both requiring a secret (shared in the hmac case), but the difference in key material for signing is significant. Verification is very different, so not trivial to extend.

This PR adds:

- Explicitly typing RSA private keys as distinct from secrets (hmac)
- Importing keys and certificates into the KeyStore
- Detecting whether to use the RSA signing flow depending on the value of ```algorithm``` field, and selecting the right hash algorithm (could be replaced on detection of supplied key?)
- Detecting whether to use the RSA verification flow depending on the type of key retrieved with ```keyId``` and use a X.509 certificate instead of a secret/privateKey.

liamdennehy committed:
RSA Signatures support, for all SHA2 variants

liamdennehy committed:
Switch for different SHA2 hash lengths

liamdennehy committed:
Only allow rsa-sha1 and rsa-256

* Specification does not permit other hash algorithms
  for RSA signatures

liamdennehy committed:
Restrict RSA algorithm selection to those in the standard

liamdennehy committed:
RSA Context tests and rename default to differentiate HMAC

aaronpk commented on src/RsaAlgorithm.php:
```diff
@@ -0,0 +1,52 @@
+          case 'sha256':
+            $algo = OPENSSL_ALGO_SHA256;
+            break;
+          case 'sha1':
```
Shouldn't this `sha1` be replaced with the 384 and 512 that are listed in `src/Algorithm.php` above?

liamdennehy replied:
Indeed, and since 384 and 512 are not listed as options in the spec I've aligned all in the PR branch

liamdennehy committed:
Resolve PHP CS Fixer issues

liamdennehy committed:
Add RSA to SignatureParametersTest

liamdennehy committed:
Add Algorithm type discriminator

* Avoid Class tests, easier to add EC keys later (also asym)

liamdennehy committed:
Better format of RSA/X509 test material

liamdennehy committed:
Generic name for key item

liamdennehy committed:
Move key retrieval to Hmac test unit

liamdennehy committed:
Extend Key class to handle RSA Keys/Certs

* Create a "Key" with only a certificate (for verification)
* Create a Key with a private key for signing/authorising
* Create a Key with both because someone will want to
* Generate exception if certificate and key are not related
* Separate signing flows for hmac (secret) and asymmetric (privateKey)

liamdennehy committed:
Separate Hmac Verifier tests

liamdennehy committed:
Verify function for RSA Algorithm

liamdennehy committed:
Use keyIds to distinguish key types in tests

liamdennehy committed:
Signing for Asymmetric keys

liamdennehy committed:
Different verification flows on key type (secret/rsa)

liamdennehy committed:
RSA Verification tests

liamdennehy committed:
Instantiate RsaContext for verify() (XDebug finding)

liamdennehy committed:
Use algorithm from instance

mtibben commented on src/Algorithm.php:
```diff
@@ -4,6 +4,9 @@
 abstract class Algorithm
 {
+    /** @var string */
+    public $type;
```
hmm what is `type` representing here? I can't see any references to "type" in the latest draft https://datatracker.ietf.org/doc/draft-cavage-http-signatures/?include_text=1

It looks like you're using this to switch behaviour - perhaps you might want to think about how to abstract this behaviour in an OO way?

liamdennehy replied:
Well that was trivially corrected. Perhaps I was over-imagining the potential flows.

Resolved in 6c7f8ed81d4a1e279488d54f2a6ba262b71c1ca7

mtibben commented on src/Algorithm.php:
```diff
@@ -18,6 +21,12 @@ public static function create($name)
         case 'hmac-sha256':
             return new HmacAlgorithm('sha256');
             break;
+        case 'rsa-sha1':
```
Should we even be supporting sha1, a very old and problematic algo?

mtibben replied:
Yeah fair enough. It might be a good idea to give a warning or even just a comment somewhere that sha1 is insecure and sha256 should be the default choice

liamdennehy replied:
I'm not a fan of the current instructions for context setup where an array of keys is supplied. We could simplify by instructing a single key be used for the signing context, possibly detecting the key type and determine the algorithm automatically - if it's not RSA or EC, it is a secret, so hmac. If RSA Private Key, RSA. By default we use SHA256. If the algorithm is explicitly specified then use the hash the user has selected.

Either that, or only select the algorithm when creating the context and set SHA256 internally by default, unless the user selects the full algorithm with hash (SHA1, SHA384, SHA512 at user's own risk for security and/or compatibility).

liamdennehy replied:
For the singing flow, the context should really only have a single key. We can derive the algorithm from that key type (actually required, can't RSA with a secret, useless to HMAC on a private key) and pick the best hash to go with it, so the user doesn't need to specify an algorithm string. Will work on it in a future PR.

liamdennehy replied:
It seems like the draft specification's authors are in favour of deprecating sha1 support:
https://github.com/w3c-dvcg/http-signatures/issues/20
We can preempt this or wait for a new revision to be published.

mtibben replied:
Yeah lets not add support for a deprecated option. Happy to leave the existing `hmac-sha1` for BC reasons

liamdennehy replied:
Well, until it's out of the spec I think it should stay in this PR to make sure we're compatible. Trivial to remove it as and when this is formally removed (and since I'm now a contributor there I'll know about it).

mtibben commented on src/Key.php:
```diff
@@ -10,13 +10,92 @@ class Key
         $this->id = $id;
-        $this->secret = $secret;
+        $certificate = Key::getX509Certificate($item);
+        $privateKey = Key::getRSAPrivateKey($item);
```
`static::getRSAPrivateKey()`?

liamdennehy replied:
It's independent of any instance state, so felt it could be static. Fixed in c2b116052048e805b6fc339cf95e80693de2eee9

mtibben commented on src/Key.php:
```diff
@@ -10,13 +10,92 @@ class Key
+            $this->secret = $item;
+            $publicKey = null;
+            $privateKey = null;
+        };
```
It looks like you're specifying a very different representation of a key here - I'd suggest the concrete types be split out and implement a similar interface

liamdennehy replied:
This took me ages to decide on. I wasn't happy just inventing a new architectural component, and burdening the user with having to understand the various key types with explicit calls. This implementation doesn't break existing flows, and automagically sorts the asymmetric types to handle the sign() and verify() flows (which don't differ in the hmac case).

Acknowledged this is not a complete implementation, but is that preferred over compatibility and user experience?

liamdennehy replied:
If we're looking at an overhaul of the Key type and KeyStore, should we incorporate #18 for the KeyStore itself? Or maybe handle the proposed work under that issue in the future and implement this solution for now?

mtibben replied:
OK so I'd suggest approaching this a little bit differently. Keep the single concrete `Key` and keep it a simple representation of the Key. Ensure your secret is a string only - mixed use variables are a pain IMO. Remove the public variables but expose a
```
public function isRSAPrivateKey()
public function isX509Certificate()
```

Then in `Verification` and `Signature` put the actual logic

liamdennehy replied:
For RSA there is no one "key", there is a signing key (private) and a verification key (public), so we need to handle these. Keep the internal representation but private, expose them via functions as you describe? Oh, and a certificate is just a container for a public key, we should use/expose it directly for consistency.

```php
public function getRSAPublicKey()
public function getRSAPrivateKey()
public function getHmacSecret()
```

For EC (in the future) add:

```php
public function getECPublicKey()
public function getECPrivateKey()
```

And with key overhaul to handle rotation in #18 (also future):

```php
public function getRSAPublicKeys($keyId)
public function getECPublicKeys($keyId)
public function getHmacSecrets($keyId)
```

liamdennehy replied:
Generalised to:
```php
public function getSigningKey()
public function getVerifyingKey()
```

Internal state for a key is now private, and the calls return the right public/private or secret for sign/verify. 2ed56e8b30405345f1c08c889490c4909ca61d3f

In future this response can be an array under a single `keyId` to handle key rotation in #18

liamdennehy replied:
@mtibben, can you review the overhaul? I think it's a lot cleaner and tolerant of different input types (e.g. array of possible verification keys or just a single key).
b9ae276fe63a6bfe3d27fa4d2eb6fa77437746fb

liamdennehy committed:
Changed static to $this methods for RSA key functions.

Resolves https://github.com/99designs/http-signatures-php/pull/35#discussion_r211120006

liamdennehy committed:
Use get_class() to discriminate instead of $type class property

liamdennehy committed:
New AlgorithmException class and common digest selection fucntion

liamdennehy committed:
New AlgorithmException class and common digest selection function

* Unit Test for bad algorithm choice
* Common algorithm selection function to improve maintainability
  when algorithms change (single implementation in verify() and sign()
* Move Algorithm instantiation to Context creation:
  Get earlier feedback on bad
  hash algorithm choice - error should occur at time of bad
  data input, not usage.

liamdennehy committed:
Verifier doesn't use a Context (it's called by a Context)

liamdennehy committed:
README with HMAC/RSA usage and separating Sign/Validate flows

* Need better documentation around key handling (e.g. no mention
  of using an existing KeyStore object)

liamdennehy committed:
Update README.md

mtibben commented on src/Context.php:
```diff
@@ -35,8 +38,8 @@ public function __construct($args)
         // algorithm for signing; not necessary for verifying.
         if (isset($args['algorithm'])) {
             $this->algorithmName = $args['algorithm'];
```
looks like this class variable is unused

liamdennehy replied:
Revoled in 0ff0bae479759100fcf0ac42bb6c58324ae6e92d

mtibben commented on src/Signature.php:
```diff
@@ -30,9 +30,21 @@ public function __construct($message, Key $key, AlgorithmInterface $algorithm, H
+            throw new Exception(
+              "Algorithm class " .
+               get_class($this->algorithm) . " unknown");
+        }
```
ideally this code wouldn't exist - use OO, not switch

liamdennehy replied:
Maybe a generic:
```php
public function getSigningKey()
public function getVirifierKey()
```

For PKI (RSA/EC), different responses, for HMAC the same object (a shared secret)?

liamdennehy replied:
Yup, above approach much simpler removing switches and requirements for key consumers to know about key details. Key state now private, keys are returned by purpose instead of explicit attributes (should make it more trivial to extend to EC too). 2bacdc7e361deeb92c97ab103a1479f8a7274445

liamdennehy committed:
Keys: Align naming of key variables with a/symmetric conventions

liamdennehy committed:
Keys: Better ways to expose Signing and Verifying keys

* Hide internal representation of keys
* Better approach for asymmetric keys in one Key object
* function getSigningKey() - Secret for HMAC, Private Key for RSA
* function getVerifyingKey() - Secret for HMAX, Piblic Key for RSA
* Move from certificates to public keys for overall consistency
* Preserve existing key import methods (may need to change with future developments)

liamdennehy committed:
Remove unused variable

liamdennehy committed:
Refer to Draft v10

* Implementation is correct against this draft, good to use
  latest link

liamdennehy commented:
Any further comments here?

mtibben commented:
Can you rebase and remove the merge commits?

mtibben commented on src/Context.php:
```diff
@@ -87,15 +89,15 @@ private function signingKey()
+    // private function algorithm()
+    // {
+    //     return Algorithm::create($this->algorithmName);
+    // }
```
remove code instead of commenting out

liamdennehy replied:
Fixed in 665baabfcb62596b838b3c852613f45b81be700e

mtibben commented on src/Key.php:
```diff
@@ -5,18 +5,131 @@
+            };
+            if ($certificate && $privateKey) {
+                // openssl_pkey_export($this->privateKey, $privateKey);
+            // openssl_x509_export($this->certificate, $certificate);
```
remove code instead of commenting out

liamdennehy replied:
Fixed in 665baabfcb62596b838b3c852613f45b81be700e

mtibben commented on src/Key.php:
```diff
@@ -5,18 +5,131 @@
+            // openssl_x509_export($this->certificate, $certificate);
+            if (! openssl_x509_check_private_key(
+                $this->certificate, $this->privateKey)
+                ) {
```
formatting is a bit off here - either inline or use consistent indenting and brackets. I've updated the php_cs_fixer on master to help with this
```php
function_name(
    $indented_arg1,
    $indented_arg2,
    $indented_arg3
);
```

mtibben replied:
but generally inlining is easiest to read

liamdennehy replied:
Fixed (Maybe?) in 665baabfcb62596b838b3c852613f45b81be700e

mtibben commented on src/Key.php:
```diff
@@ -5,18 +5,131 @@
+                $privateKey = openssl_get_privatekey($object);
+            } catch (\Exception $e) {
+                $privateKey = null;
+            };
```
Any reason we want to catch exceptions here? And if so can we be more explicit about what exception?

mtibben commented on src/RsaAlgorithm.php:
```diff
@@ -0,0 +1,60 @@
+    {
+        $algo = $this->getRsaHashAlgo($this->digestName);
+        if (! openssl_get_privatekey($signingKey)) {
+            throw new httpSignatures\AlgorithmException("OpenSSL doesn't understand the supplied key (not valid or not found)");
```
capital H

mtibben commented on src/Verification.php:
```diff
@@ -39,8 +39,28 @@ public function isValid()
+                $expectedResult = hash_hmac(
+                  'sha256', $this->expectedSignatureBase64(), $random, true);
+                $providedResult = hash_hmac(
+                  'sha256', $this->providedSignatureBase64(), $random, true);
```
source formatting looks off here too - 4 space indentation, and inline would be easier to read here

liamdennehy replied:
Fixed in 665baabfcb62596b838b3c852613f45b81be700e

liamdennehy committed:
PHP Code Style Fixups

liamdennehy committed:
Remove comments, consistent indentation, better function styles

liamdennehy committed:
Typos & Code Style

mtibben commented:
Hey @liamdennehy I'm not sure you're following my request regarding the rebase, so will try be more clear :) We follow a "feature branch" or "[Github Flow](https://guides.github.com/introduction/flow/)" workflow. So:

1. We don't want merge commits. [Rebase them out](https://superuser.com/questions/648281/remove-a-merge-commit-in-git)
2. We want your branch to be from the most recent master commit. [Rebase on master](https://stackoverflow.com/questions/7929369/how-to-rebase-local-branch-with-remote-master/18442755#18442755)

liamdennehy commented:
@mtibben There's a strong possibility I have no idea what I'm doing ☺️
Thanks for the clarification.

mtibben commented on src/Context.php:
```diff
@@ -16,6 +16,9 @@ class Context
     private $signingKeyId;
 
+    /** @var AlgorithmInterface */
+    public $algorithm;
```
private

liamdennehy replied:
Resolved in 1143d2a9e043b91846ad760f21c3b15f1fc85750

mtibben commented on src/Context.php:
```diff
@@ -34,9 +37,8 @@ public function __construct($args)
         // algorithm for signing; not necessary for verifying.
         if (isset($args['algorithm'])) {
-            $this->algorithmName = $args['algorithm'];
+            $this->algorithm = Algorithm::create($args['algorithm']);
```
Any reason for this change?

liamdennehy replied:
This old parameter only handled the HMAC case, where a singing and verifying key are the same. Needed a class to discriminate HMAC from RSA, so invoke with a factory that makes the right underlying Algorithm subclass.

mtibben commented on src/Key.php:
```diff
@@ -5,18 +5,132 @@
-        $this->secret = $secret;
+        $certificate = $this->getX509Certificate($item);
+        $privateKey = $this->getRSAPrivateKey($item);
+        if (($certificate || $privateKey)) {
```
this if statement seems unnecessary as your nested `if` check the same thing. Try to avoid nesting control statements

liamdennehy replied:
HMAC only takes a single key for signing and verifying, RSA/EC require two distinct keys, so this approach allows the existing Key class and KeyStore interface to be used for both since we don't have an explicit method for supplying one or the other. Alternative is to break Key into different subtypes (e.g. HMAC & PKI or Shared, Private & Public), but this flow is tolerant of whatever the user throws into the input array - this could be a private key, a certificate, or both.

mtibben commented on src/Key.php:
```diff
@@ -5,18 +5,132 @@
+            if ($certificate && $privateKey) {
+                if (!openssl_x509_check_private_key(
+                        $this->certificate,
+                        $this->privateKey
```
code style - just inline it, there's no benefit for breaking this up

liamdennehy replied:
I'm trying to keep the width below 80 chars, and for me at least breaking this up helps me visualise the logic blocks.

mtibben commented on src/Key.php:
```diff
@@ -5,18 +5,132 @@
+
+    private function getRSAPrivateKey($object)
+    {
+        if (is_array($object)) {
```
Under what circumstances is this going to be an array and why? Ideally avoid allowing multiple types in arguments

liamdennehy replied:
The user either supplies:
- A Public Key/Certificate for verifying as string/object
- A Private Key for signing as string/object
- Or both together in an array of string/objects - this can now be used for both signing and verifying without having to construct different objects for each case.

Allowing both forms means we can use the same KeyStore object for all three cases to derive a Key object. Alternative is to expose HMAC and Asymmetric-specific Key interfaces which is both a BC and requires more documentation and user education for little gain.

liamdennehy replied:
@mtibben Is this ok? I realise the input here is fungible, but I'd rather that than expect the user to do key/certificate handling to a more advanced degree.

mtibben commented on src/Key.php:
```diff
@@ -5,18 +5,132 @@
+    private function getX509Certificate($object)
+    {
+        $key = null;
+        if (is_array($object)) {
```
same here

mtibben commented on src/Key.php:
```diff
@@ -5,18 +5,132 @@
+                $result = openssl_get_publickey($object);
+            } catch (\Exception $e) {
+                $result = null;
+            }
```
why are you try / catching ?

liamdennehy replied:
We're not specifying which order the public/private keys should be provided, and no specific Key subclasses for each, so have to detect:
- If the result is not a public key, ignore it - the same object could resolve to a private key which is useless for verification.
- Same for openssl_get_privatekey() test.

Will see if there's a non-exception based detection mechanism, but the openssl libraries aren't very friendly.

liamdennehy committed:
Private variable, remove `@throws` hints

liamdennehy commented:
@mtibben I'm going to overhaul the Key class to make the flow easier to track, eg
``` php
    static function hasX509Certificate($object)
```
...to explicitly test if an object has a certificate instead of try..catch blocks all over the place.

This will also make a good base to extend the class for #18 (Key Rotation). Will ping you when it's ready.

liamdennehy committed:
Cleaner Key class and sundry names

* Key now handles inputs as X.509, naked Public Key and
  Private Key (all PEM)
* Improved & clearer detection logic with explicit tests
  for asymmetric key material, falling back on shared secret
  for HMAC if asymmetric keys are not detected
  hasX509Certificate()
  hasPublicKey()
  hasPrivateKey()
* Clearer names for test constants (easier to expand to EC later)

liamdennehy committed:
phpdoc annotations

liamdennehy commented:
Any further questions, or updates on open conversations in this PR?

lahaxearnaud commented:
Hello,

Can you please merge and tag this PR ?

Regards,

Arnaud

liamdennehy commented:
Apologies, accidentally delete the branch for this PR in my source.

However, I have integrated this feature in my own project along with a number of others, and published this in [packagist](https://packagist.org/packages/liamdennehy/http-signatures-php). Documentation for the entire library is published at Read the Docs: [http-signatures-php](https://http-signatures-php.readthedocs.io/en/stable/) - incomplete but being expanded regularly.

liamdennehy committed:
Travis build fixes and include phpunit script

- Remove unsupported Symfony/PHP version combinations
- Add PHP 7.3 support

liamdennehy commented:
Closing PR due to inactivity.
